src/core/Rule.ml

```diff
@@ -28,7 +28,46 @@ open Ppx_hash_lib.Std.Hash.Builtin
  * See also Mini_rule.ml where formula and many other features disappear.
  *)
 
-type rule_id = string [@@deriving show]
+(*
+   A rule ID is a string. This creates a dedicated type to clarify interfaces
+   and error messages. The coercion operator can be used as an alternative
+   to 'to_string x': '(x :> string)' to obtain a plain string.
+*)
+module ID : sig
+  type t = private string [@@deriving show, eq]
+
+  val to_string : t -> string
+  val of_string : string -> t
+  val to_string_list : t list -> string list
+  val of_string_list : string list -> t list
+  val compare : t -> t -> int
+
+  (*
+     Rule ids are prepended with the `path.to.the.rules.file.`, so
+     when comparing a rule (r) with the rule to be included or excluded,
+     allow for a preceding path
+
+       "path.to.foo.bar" ~suffix:"foo.bar" -> true
+       "foo.bar" ~suffix:"foo.bar" -> true
+       "xfoo.bar" ~suffix:"foo.bar" -> false
+
+     TODO we may want to allow globs for rules
+  *)
+  val ends_with : t -> suffix:t -> bool
+end = struct
+  type t = string [@@deriving show, eq]
+
+  let to_string x = x
+  let of_string x = x
+  let to_string_list x = x
+  let of_string_list x = x
+  let compare = String.compare
+
+  let ends_with r ~suffix:inc_or_exc_rule =
+    r = inc_or_exc_rule || String.ends_with ~suffix:("." ^ inc_or_exc_rule) r
+end
+
+type rule_id = ID.t [@@deriving show, eq]
```

Comment:
We could move this ID module to its own file if people prefer it that way.

Reply:
It's fine